packages/runtime/container-runtime/src/summarizer.ts

```diff
@@ -277,11 +276,8 @@ export class Summarizer extends EventEmitter implements ISummarizer {
         });
 
         // Listen for ops
-        this.systemOpListener = (op: ISequencedDocumentMessage) => runningSummarizer.handleSystemOp(op);
-        this.runtime.deltaManager.inbound.on("op", this.systemOpListener);
-
-        this.opListener = (error: any, op: ISequencedDocumentMessage) => runningSummarizer.handleOp(error, op);
-        this.runtime.on("batchEnd", this.opListener);
+        this.opListener = (op: ISequencedDocumentMessage) => runningSummarizer.handleOp(op);
```

Comment:
@vladsud I thought about how we may move this handler logic to before we construct the object, but I don't think I like the idea. It would have to be very reliant on how this `Summarizer.start(...)` method and further up is called (because of asynchronous nature), so it may make a lot more sense to have this backtracking logic handled by the `RunningSummarizer` after we call `waitStart`.

For example:
- We initialize the `RunningSummarizer` object and wait for it to start
- Provide the `RunningSummarizer` some more up-to-date op sequence number and pass that along to the heuristic data appropriately
- Tell the heuristic runner to start and also run an initial heuristic check

Is there somewhere we can get an up-to-date op sequence number easily? Or will we have to do implement something new to track this and then pass the information along to the `RunningSummarizer`?

Reply:
It's trivial to get number of ops (as we know latest op we processed, and reference sequence number of a summary we loaded from). That's what existing code (before your changes) does when deciding when to run a summary.
We will lose break down of info on user vs. system ops though in that way. It may be fine, as we can guess / do some allocation, i.e. assume that it's 50% - 50% or some other rate that makes sense.

Reply:
Yes, we'd probably have to assume some basic op ratio or potentially do a different check altogether. But currently we only update the current op sequence number in the op handler for the `RunningSummarizer`. My understanding is that there may be some ops we'd like to process a summary for in the following cases:
1. Op is processed during the time that we are initiating the `RunningSummarizer` instance but before we listen for the op events (will get missed by the handlers in the current workflow)
2. Op was sequenced after the last time we summarized (op sequence number > summarize ref sequence number)

So, is there currently a good way to retrieve the most recent op sequence number (so we can run heuristics) before we update the value for `this.heuristicData.lastOpSequenceNumber`?

Reply:
lastOpSequenceNumber looks very weird.
Yes, you can always get it from IDeltaManager.lastSequenceNumber, which is same as IDeltaManager.lastMessage.sequenceNumber.

Summarizer has access to delta manager through this.runtime.deltaManager, I have not looked if it's propagated through the layers.

You actually see it used here, where we populate initial value:
```ts
        const runningSummarizer = await RunningSummarizer.start(
            this.logger,
            this.summaryCollection.createWatcher(clientId),
            this.configurationGetter(),
            async (...args) => this.internalsProvider.submitSummary(...args), // submitSummaryCallback
            new SummarizeHeuristicData(
                this.runtime.deltaManager.lastSequenceNumber,
                { /** summary attempt baseline for heuristics */
                    refSequenceNumber: this.runtime.deltaManager.initialSequenceNumber,
                    summaryTime: Date.now(),
                } as const,
            ),
```

Are you sure there is no way to move initialization of handler up in time? :)
I think you will need to pass runtime down to RunningSummarizer either way, and if so, we can move responsibility to register for events into RunningSummarizer. I think it's more natural flow - it feels weird that Summarizer.start() does this event registration for RunningSummarizer